### PR TITLE
Avoid resetting the VF if the netns does not exist 

### PR DESCRIPTION
**Pull request by zeeke**
The VF might have been assigned to another running
Pod if the network namespace is not present. In cases
like this, resetting the VF might break the configuration
of the running Pod.
The above scenario might occur when the IPAM plugin takes a lot of time to complete, and the CmdDel gets called
multiple times.

Check if the namespace is still present before resetting the VF

@bn222 @SchSeba WDYT about this approach?

### CODE REVIEW

**Review by e0ne — APPROVED** [review]

File: cmd/sriov/main.go [hunk 1]
```diff
@@ -270,16 +270,10 @@ func cmdDel(args *skel.CmdArgs) error {
...
-	   reset netdev VF with trust off. So, reset VF MAC address via PF first.
-	*/
-	if err := sm.ResetVFConfig(netConf); err != nil {
-		return fmt.Errorf("cmdDel() error reseting VF: %q", err)
-	}
+	var netns ns.NetNS
 
 	if !netConf.DPDKMode {
-		netns, err := ns.GetNS(args.Netns)
+		netns, err = ns.GetNS(args.Netns)
```
Comment (SchSeba) [inline cmd/sriov/main.go:276]: I think we should always do this not only for non dpdk devices.
if not we should be able to reproduce the issue on a vfio-pci device for example (maybe not using the ipam)
but for example if the driver is slow and we need to reset the vlan on the vf and that part takes a lot of time we maybe be able to reproduce the issue no?

Comment (zeeke) [reply cmd/sriov/main.go]: that's a good point. I updated the code so that if the net namespace doesn't exist, then Release and Reset are not called.
WDYT?

**Review by SchSeba — APPROVED** [review]
This LGTM! nice work!

**Comment by SchSeba** [comment]
@adrianchiris can you please give this one a look so we can merge it?

File: cmd/sriov/main.go [hunk 2]
```diff
@@ -270,6 +270,29 @@ func cmdDel(args *skel.CmdArgs) error {
...
+		// if provided path does not exist (e.x. when node was restarted)
+		// plugin should silently return with success after releasing
+		// IPAM resources
+		_, ok := err.(ns.NSPathNotExistErr)
+		if ok {
+			logging.Debug("Exiting as the network namespace does not exists anymore",
+				"func", "cmdDel",
+				"netConf.DeviceID", netConf.DeviceID,
+				"args.Netns", args.Netns)
+			return nil
```
Comment (adrianchiris) [inline cmd/sriov/main.go:286]: in this case dont we want to skip to L315 and mark pci address as released ?

**Comment by adrianchiris** [comment]
overall looks good, nasty bug :)

File: cmd/sriov/main.go (hunk 2 above)
Comment (zeeke) [reply cmd/sriov/main.go:286]: I think it's better not:

The scenario I'm trying to solve here is when a CmdDelete is called when the device has already been allocated to another Pod. If we call `DeleteAllocatedPCI(...)`, we are freeing up a device on a running pod.

WDYT?

Comment (adrianchiris) [reply cmd/sriov/main.go:286]: can this really happen ? we use the pci allocator to check if the device is allocated in CmdAdd, see [1]

so even if NS was deleted and device was re-assigned to a new pod, its CNI call would fail on ADD until we released the device.

[1] https://github.com/k8snetworkplumbingwg/sriov-cni/blob/fca6591297c0e52b6522573bd367ccb7d6551fd0/pkg/config/config.go#L60

Comment (zeeke) [reply cmd/sriov/main.go:286]: You're right; in the scenario I just described, it probably can't happen.
What we saw happening is related to a very long execution of [`ipam.ExecDel(...)`](https://github.com/k8snetworkplumbingwg/sriov-cni/pull/313/files#diff-30bddd0b0a3414968a5fa92d19e50ad3168e273728fd52ba867afbc442defe62R255).
This scenario includes 3 sriov-cni calls:
a. **first_cniDEL** starts and call ipam.ExecDel(...). this is going to take a very long time (~2 minutes)
b. **second_cniDEL** is invoked and, for some reason, it completes quite quickly (ipam deletion, vlan/mac reset, ...). PIC is deallocated, netns is deleted.
c. **cniADD** is  invoked. It assigns the VF to a new pod, configuring VLAN, MAC, ...
d. **first_cniDEL** continues its execution, as the `ipam.ExecDel` finished. Here, what I want to avoid is resetting the VF MAC, VLAN and so on.

Comment (adrianchiris) [reply cmd/sriov/main.go:286]: maybe a better solution is to use lock file to handle on a single cmd ADD/DEL at a time for a given ns, network, interface ?
that way second cni DEL will fail and prevent freeing up the PCI in the allocator ?

Comment (SchSeba) [reply cmd/sriov/main.go:286]: I was thinking the same but I don't want to have another file maybe we can use the existing PCI file and have the lock on that file?

Comment (adrianchiris) [reply cmd/sriov/main.go:286]: we can flock the pci file if it works. just need to make sure we release when we exit ADD and DEL cmds.

Comment (adrianchiris) [reply cmd/sriov/main.go:286]: looking at the code, we save PCI device at the end of the cmd add call.
in this case we want to "lock" the call for the ns/network/interface at the beginning.
so if we want to use the same file we need to be careful (save at the beginning and release in case of failure)

Comment (SchSeba) [reply cmd/sriov/main.go:286]: you are right so maybe having a tmp file for `ns/network/interface` is the right way going forward but we must be sure we clean that file.

and we cover a case of unexpected reboot so maybe putting this lock file in /tmp/ so on reboot it will get clean?

Comment (adrianchiris) [reply cmd/sriov/main.go:286]: sounds good.

Comment (adrianchiris) [reply cmd/sriov/main.go:286]: > we can flock the pci file if it works. just need to make sure we release when we exit ADD and DEL cmds.

so , according to flock docs[1] it would release the lock if the process exits so in case of reboot we would not be stuck in a deadlock (it would be released when process exits)

[1] https://man7.org/linux/man-pages/man2/flock.2.html

Comment (zeeke) [reply cmd/sriov/main.go:286]: I'm trying to create a reproducer (not easy at all!) and see if the flock solves the problem. 

> instead of /tmp, we should go through all the files we have put on the fs when we start up the operator controller, and ensure our local db (files we put on the file system) is clean. There is no guarantee that /tmp will be cleared.

This is a good point. We might move the sriov-cni operation folder from `/var/lib/cni/sriov` to `/run/sriov`. That folder is supposed to be clean at every boot.

https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s15.html

Comment (bn222) [reply cmd/sriov/main.go:286]: I like how you say "is supposed to be". I would like to build code that does not have "supposed to be". If you do, and the assumption is wrong, then you have a bug. Better to not assume anything and build robust code that checks the system state instead of comparing to some cached version of the state stored on a file system

**Comment by zeeke** [comment]
I implemented the fix via the `unix.Flock` strategy. 

I manually tested these changes on OpenShift 4.17 and with automated tests (See #318)

Please take another look

**Review by SchSeba — APPROVED** [review]
LGTM!

Nice work

**Comment by zeeke** [comment]
@adrianchiris @e0ne Please take a look. This PR fixes an important scenario for OpenShift

File: pkg/utils/pci_allocator.go [hunk 3]
```diff
@@ -25,6 +29,40 @@ func NewPCIAllocator(dataDir string) *PCIAllocator {
 	return &PCIAllocator{dataDir: filepath.Join(dataDir, "pci")}
 }
 
+// Lock gets an exclusive lock on the given PCI address, ensuring there is no other process configuring / or de-configuring the same device.
+func (p *PCIAllocator) Lock(pciAddress string) error {
+	if err := os.MkdirAll(p.dataDir, 0600); err != nil {
+		return fmt.Errorf("failed to create the sriov data directory(%q): %v", p.dataDir, err)
+	}
+
+	path := filepath.Join(p.dataDir, pciAddress)
```
Comment (adrianchiris) [inline pkg/utils/pci_allocator.go:38]: i think the file might exist,  if you lock but later on fail CMD add call (and saveAllocatedPCI was not called).
i dont think anything bad will happen since on del we wont have any cached config and return early but still this file will be left.

next time cmd add is called with same address it will hit the error at line 115.

alternative: save new file  with .lock prefix e.g "0000:03:00.4.lock"

then have an unlock function which will delete the file if exists ?
(called via defer on main methods i.e add/del after we gained lock)

that said im fine with current way as well. so i leave it up 2 u.

**Review by adrianchiris — APPROVED** [review]
please see comment

File: pkg/utils/pci_allocator.go (hunk 3 above)
Comment (zeeke) [reply pkg/utils/pci_allocator.go]: I added the `.lock` suffix to the flock files, but cleaning them up on exit is not very straightforward.
TL;DR I prefer to leave them in the filesystem, they would not hurt any other CNI operation

Consider the following scenario:
```
process1 flocks file1 using file descriptor f1
process2 tries to lock file1 and waits. file descriptor is f2
process1 deletes file1 and exits
process2 continues its work
process3 starts and flocks file1, using file descriptor fd3.

fd3 and fd2 are not related, so process 3 acquires the lock
```

The following script proves it:
```
flock -x file1 sh -c "echo `date`' p1 starts'; sleep  2; rm file1; echo `date`' p1 ends'" &
sleep 0.1
flock -x file1 sh -c "echo `date`' p2 starts'; sleep 10;           echo `date`' p2 ends'" &
sleep 3
flock -x file1 sh -c "echo `date`' p3 starts'; sleep  1;           echo `date`' p3 ends'" &
wait
```
execution:
```
Thu Feb 20 06:52:09 PM CET 2025 p1 starts
Thu Feb 20 06:52:09 PM CET 2025 p1 ends
Thu Feb 20 06:52:09 PM CET 2025 p2 starts
Thu Feb 20 06:52:12 PM CET 2025 p3 starts
Thu Feb 20 06:52:12 PM CET 2025 p3 ends
Thu Feb 20 06:52:09 PM CET 2025 p2 ends
```

File: pkg/utils/pci_allocator.go [hunk 4]
```diff
@@ -25,6 +29,40 @@ func NewPCIAllocator(dataDir string) *PCIAllocator {
...
+	if err := os.MkdirAll(p.dataDir, 0600); err != nil {
+		return fmt.Errorf("failed to create the sriov data directory(%q): %v", p.dataDir, err)
+	}
+
+	path := filepath.Join(p.dataDir, pciAddress+".lock")
+
+	// unix.O_CREAT - Create the file if it doesn't exist
+	// unix.O_RDWR - Open the file for read/write
+	// unix.O_CLOEXEC - Automatically close the file on exit. This is useful to keep the flock until the process exits
+	fd, err := unix.Open(path, unix.O_CREAT|unix.O_RDWR|unix.O_CLOEXEC, 0600)
```
Comment (adrianchiris) [inline pkg/utils/pci_allocator.go:43]: FYI: https://github.com/gofrs/flock   we use it in ib-sriov-cni

Comment (adrianchiris) [inline pkg/utils/pci_allocator.go:43]: nit: open with read-only and not read-write.

File: pkg/utils/pci_allocator.go (hunk 3 above)
Comment (adrianchiris) [reply pkg/utils/pci_allocator.go]: ack, im ok with leaving them.

from tidiness perspective, WDYT about putting the lockfiles under their own folder in data dir ?

**Review by SchSeba — APPROVED** [review]

File: pkg/utils/pci_allocator.go (hunk 4 above)
Comment (zeeke) [reply pkg/utils/pci_allocator.go:43]: I saw it, but for this case, I prefer to go directly with the syscalls. 
The wrapper looks bigger than the wrapped logic

Comment (zeeke) [reply pkg/utils/pci_allocator.go:43]: fixing

**Comment by zeeke** [comment]
> ack, im ok with leaving them.
> from tidiness perspective, WDYT about putting the lockfiles under their own folder in data dir ?

Makes sense. Changing

**Comment by zeeke** [comment]
> just two small nits and a question after the program existed will the file remain on the host? meaning are we going to have left overs of .lock files on the node? if that is the case I think we should clean them no?

That's a good point but it is not that easy. See this comment:
https://github.com/k8snetworkplumbingwg/sriov-cni/pull/313#discussion_r1964091596

I fixed the other comments

**Review by SchSeba — APPROVED** [review]
LGTM
